llama-index-core/llama_index/core/chat_engine/condense_plus_context.py

```diff
@@ -150,41 +183,104 @@ async def _acondense_question(
                 nodes, query_bundle=QueryBundle(message)
             )
 
-        context_str = "\n\n".join(
-            [n.node.get_content(metadata_mode=MetadataMode.LLM).strip() for n in nodes]
-        )
-        return context_str, nodes
+        return nodes
 
-    async def _aretrieve_context(self, message: str) -> Tuple[str, List[NodeWithScore]]:
-        """Build context for a message from retriever."""
+    async def _aget_nodes(self, message: str) -> List[NodeWithScore]:
+        """Generate context information from a message."""
         nodes = await self._retriever.aretrieve(message)
         for postprocessor in self._node_postprocessors:
             nodes = postprocessor.postprocess_nodes(
                 nodes, query_bundle=QueryBundle(message)
             )
 
-        context_str = "\n\n".join(
-            [n.node.get_content(metadata_mode=MetadataMode.LLM).strip() for n in nodes]
+        return nodes
+
+    def _get_prefix_qa_messages_with_context(
+        self, chat_history: List[ChatMessage]
+    ) -> List[ChatMessage]:
+        """Get the prefix messages with context."""
+        system_prompt = self._system_prompt or ""
+        context_str_w_sys_prompt = self._context_prompt_template + system_prompt.strip()
+        return [
+            ChatMessage(
+                content=context_str_w_sys_prompt, role=self._llm.metadata.system_role
+            ),
+            *chat_history,
+            ChatMessage(content="{query_str}", role=MessageRole.USER),
+        ]
+
+    def _get_prefix_refine_messages_with_context(
+        self, chat_history: List[ChatMessage]
+    ) -> List[ChatMessage]:
+        """Get the prefix messages with context."""
+        system_prompt = self._system_prompt or ""
+        context_str_w_sys_prompt = (
+            self._context_refine_prompt_template + system_prompt.strip()
+        )
+
+        return [
+            ChatMessage(
+                content=context_str_w_sys_prompt, role=self._llm.metadata.system_role
+            ),
+            *chat_history,
+            ChatMessage(content="{query_str}", role=MessageRole.USER),
+        ]
+
+    def _get_response_synthesizer(
+        self, chat_history: List[ChatMessage], streaming: bool = False
+    ) -> CompactAndRefine:
+        qa_messages = self._get_prefix_qa_messages_with_context(chat_history)
+        refine_messages = self._get_prefix_refine_messages_with_context(chat_history)
+        return self._synthesizer_cls(
+            llm=self._llm,
+            callback_manager=self.callback_manager,
+            text_qa_template=ChatPromptTemplate.from_messages(qa_messages),
+            refine_template=ChatPromptTemplate.from_messages(refine_messages),
+            streaming=streaming,
+        )
+
+    def _get_current_chat_history(
+        self, input_msg: str, nodes: List[NodeWithScore]
+    ) -> List[ChatMessage]:
+        # get the current memory buffer safely
+        if hasattr(self._memory, "tokenizer_fn"):
```

Comment:
what is this for again

Reply:
also i assume the memory doesn't include the system prompt?

Reply:
It's so that memory is retrieved taking the size of the retrieved context into account -- debatable if this is still needed or not

Reply:
actually i'm not sure i understand, can you ELI5 why we limit memory based on retrieved context size

Reply:
So
- the user sets some token limit on the memory
- they would expect this to limit the input size of the memory buffer
- BUT the retrieved context is also added to the current memory buffer
- so, it will end up exceeding the token limit that you set
- a contributor added this functionality to make it possible to account for the retrieved context
- I carried this over from the previous version of the chat engine

Reply:
eh, just pulled out this concept, it was mostly a workaround to not using synthesizers before